lib/rspec/mocks/instance_method_stasher.rb

```diff
@@ -139,7 +139,12 @@ def method_owned_by_klass?
         # The owner of M.b is the raw Module object, instead of the expected
         # singleton class of the module
         return true if RUBY_VERSION < '1.9' && owner == @object
-        owner == @klass || !(method_defined_on_klass?(owner))
+
+        owner == @klass ||
+          # When `extend self` is used, and not under any instance of
+          (owner.singleton_class == @klass &&
```

Comment:
I think it's ok to use `singleton_class` now because we can drop Ruby 1.8.7 and all later Rubies do support it.

I can't get Ruby 1.8.7 to work anyway... even with `class << owner; self; end` there are still other wild failures in Ruby 1.8.7.

Reply:
> I think it's ok to use singleton_class now because we can drop Ruby 1.8.7

Ruby 1.8.7 support will not be dropped until RSpec 4.

Reply:
> Ruby 1.8.7 support will not be dropped until RSpec 4.

That is fine as long as it's scheduled. Before then we'll just pick this patch into our codebase.

Reply:
Just be aware that this won't be merged at this time without Ruby 1.8.7 support.